cupy/testing/parameterized.py

```diff
@@ -32,68 +88,113 @@ def new_v(self, *args, **kwargs):
-        cls, predicate=lambda _: inspect.ismethod(_) or inspect.isfunction(_))
-    for name, method in members:
-        if name.startswith('test_'):
-            setattr(cls, name, wrap_test_method(method))
-
-    # Add new test class to module
-    setattr(module, cls_name, cls)
-
-
-def _gen_cases(name, base, params):
-    module = sys.modules[name]
-    for i, param in enumerate(params):
-        _gen_case(base, module, i, param)
+                _raise_from(e.__class__, s.getvalue(), e)
+        return new_method
+
+    return (cls_name, mb, method_generator)
 
 
 def parameterize(*params):
-    def f(klass):
-        assert issubclass(klass, unittest.TestCase)
-        _gen_cases(klass.__module__, klass, params)
-        # Remove original base class
-        return None
-    return f
+    # TODO(niboshi): Add documentation
+    return _bundle.make_decorator(
+        lambda base: _parameterize_test_case_generator(base, params))
+
+
+def _values_to_dicts(names, values):
+    assert isinstance(names, six.string_types)
+    assert isinstance(values, (tuple, list))
+
+    def safe_zip(ns, vs):
+        if len(ns) == 1:
+            return [(ns[0], vs)]
+        assert isinstance(vs, (tuple, list)) and len(ns) == len(vs)
+        return zip(ns, vs)
+
+    names = names.split(',')
+    params = [dict(safe_zip(names, value_list)) for value_list in values]
+    return params
+
+
+def from_pytest_parameterize(names, values):
+    # Pytest-style parameterization.
+    # TODO(niboshi): Add documentation
+    return _values_to_dicts(names, values)
+
+
+def parameterize_pytest(names, values):
+    # Pytest-style parameterization.
+    # TODO(niboshi): Add documentation
+    return parameterize(*from_pytest_parameterize(names, values))
 
 
 def product(parameter):
-    keys = sorted(parameter)
-    values = [parameter[key] for key in keys]
-    values_product = itertools.product(*values)
-    return [dict(zip(keys, vals)) for vals in values_product]
+    # TODO(niboshi): Add documentation
+    if isinstance(parameter, dict):
+        return product_dict(*[
+            _values_to_dicts(names, values)
+            for names, values in sorted(parameter.items())])
+
+    elif isinstance(parameter, list):
+        # list of lists of dicts
+        if not all(isinstance(_, list) for _ in parameter):
+            raise TypeError('parameter must be list of lists of dicts')
+        if not all(isinstance(_, dict) for l in parameter for _ in l):
+            raise TypeError('parameter must be list of lists of dicts')
+        return product_dict(*parameter)
+
+    else:
+        raise TypeError(
+            'parameter must be either dict or list. Actual: {}'.format(
+                type(parameter)))
 
 
 def product_dict(*parameters):
+    # TODO(niboshi): Add documentation
     return [
         {k: v for dic in dicts for k, v in six.iteritems(dic)}
         for dicts in itertools.product(*parameters)]
+
+
+def _nd_indices_to_cover_each_2d(
```

Comment:
I suppose that this function can be also removed.

Reply:
yes, thanks. I have updated the last commit to remove it